src/relay/backend/graph_executor_codegen.cc

```diff
@@ -189,9 +192,109 @@ class GraphExecutorCodegen : public backend::MemoizedExprTranslator<std::vector<
     targets_ = targets;
   }
 
+  /*!
+   * \brief Calculate the storage required to store the type of relay.Expr
+   *
+   * \param func The relay expr for which the storage is calculated
+   */
+  int64_t CalculateRelayExprSizeBytes(const Type& expr_type) {
+    if (expr_type->IsInstance<TupleTypeNode>()) {
+      auto tuple_type = Downcast<TupleType>(expr_type);
+      int64_t size = 0;
+      for (const auto& field : tuple_type->fields) {
+        size += CalculateRelayExprSizeBytes(field);
+      }
+      return size;
+    }
+    auto tensor_type = expr_type.as<TensorTypeNode>();
+    auto shape = tensor_type->shape;
+    int num_of_elements = 1;
+    for (const auto& dim_index_expr : shape) {
+      if (dim_index_expr->IsInstance<IntImmNode>()) {
+        num_of_elements *= dim_index_expr.as<IntImmNode>()->value;
+      } else {
+        // If shape is dynamic, we cannot calculate workspace in compile time.
+        num_of_elements = 0;
+      }
+    }
+    auto element_size = tensor_type->dtype.bytes();
+    return element_size * num_of_elements;
+  }
+
+  /*!
+   * \brief Update the "main" control function's metadata
+   *
+   * \param func The main function that contains calls to relay primitive functions
+   */
+  void UpdateMainWorkspaceSize(const Function& func) {
+    std::unordered_map<int, std::unordered_map<int, int>> sid_workspace;
+    std::unordered_map<int, int> device_workspace;
```

Comment:
would suggest to move this to line 268, where it's built up, so that it's clear it's not needed til then.

Reply:
I did a different change -- kind of unified the inits. See if you agree.

Reply:
i think better is moving the variable declarations closer to where they're used. it makes it clear that device_workspace isn't of concern when analyzing the first code stanza.

Reply:
Alright; I moved it down